### PR TITLE
lgd_users

### PR DESCRIPTION
**Pull request by kellylanjaya**

### CODE REVIEW

**Review by toileto — CHANGES_REQUESTED** [review]
@kellylanjaya, overall sudah oke, cuma ada beberapa point improvement

1. Yang `.DS_Store` nya di hapusin dulu yaaa 😁 
2. Jangan lupa kalau udah mau release, table nya ganti di FROM nya buat ke `prod`
3. Ketimbang camelCase, pakainya snake_case yaaaa

See comments untuk point 2 & 3 for further details yaaaa 🙇

**Review by ameliaksalwa — CHANGES_REQUESTED** [review]
crosscheck yaa

**Review by fikrihaykal41 — APPROVED** [review]
sejauh yg ku lihat sudah oke, aku sih yes

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.subDistrictId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.subDistrictId'), '"', ''))  AS sub_district_id,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.zipCode'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.zipCode'), '"', ''))  AS zip_code,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.zipCodeId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.zipCodeId'), '"', ''))  AS zipcode_id,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.address'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.address'), '"', ''))  AS address,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.mainAddress'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.mainAddress'), '"', ''))  AS main_address,
+      IF(JSON_EXTRACT_SCALAR(list_address, '$.externalId') = "", NULL, JSON_EXTRACT(list_address, '$.externalId'))  AS external_id,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.lat'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.lat'), '"', ''))  AS lat,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:89]: ```IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.lat'), '"', '') = "", NULL, CAST(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.lat'), '"', '') AS FLOAT64)```

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+      IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.long'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.long'), '"', ''))  AS long,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:90]: ```IF(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.long'), '"', '') = "", NULL, CAST(REPLACE(JSON_EXTRACT_SCALAR(list_address, '$.long'), '"', '') AS FLOAT64) AS long,```

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+-- BEGIN MAIN_ADDRESS
+,main_address AS (
+  SELECT
+    data,
+    ts AS published_timestamp,
+    STRUCT(
+      IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.mainAddress.storeName'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.mainAddress.storeName'), '"', ''))  AS store_name,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.mainAddress.recipientName'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.mainAddress.recipientName'), '"', ''))  AS recipient_name,
+      CAST(JSON_EXTRACT_SCALAR(data, '$.usermetadata.mainAddress.phoneNumber') AS INT64) AS phone_number,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:129]: `STRING` yaaaa phone_number

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+-- END MAIN_ADDRESS
+
+-- BEGIN STORE_OWNER
+,store_owner AS (
+  SELECT
+    data,
+    ts AS published_timestamp,
+    STRUCT (
+      IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerName'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerName'), '"', ''))  AS owner_name,
+      CAST(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerPhoneNumber'), '"', '') AS INT64) AS owner_phone_number,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:158]: Ini jug jadikan ke `STRING` yaaa

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.subDistrictId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.subDistrictId'), '"', ''))  AS sub_district_id,
+        IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.zipCode'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.zipCode'), '"', ''))  AS zip_code,
+        IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.zipCodeId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.zipCodeId'), '"', ''))  AS zip_code_id,
+        IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.address'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.address'), '"', ''))  AS address,
+        IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.lat'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.lat'), '"', ''))  AS lat,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:171]: ```IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.lat'), '"', '') = "", NULL, CAST(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.lat'), '"', '') AS FLOAT64)) AS lat```

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.long'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.long'), '"', ''))  AS long
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:172]: ```IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.long'), '"', '') = "", NULL, CAST(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.storeOwner.ownerAddress.long'), '"', '') AS FLOAT64)```

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+-- BEGIN INTEGRATED_ACCOUNT
+,integrated_account AS (
+  SELECT
+    data,
+    ts AS published_timestamp,
+    STRUCT(
+      STRUCT (
+      IF(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.integratedAccount.tmoney.userName'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.integratedAccount.tmoney.userName'), '"', ''))  AS user_name,
+      CAST(REPLACE(JSON_EXTRACT_SCALAR(data, '$.usermetadata.integratedAccount.tmoney.phoneNo'), '"', '') AS INT64) AS phone_number,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:188]: Ke `STRING` yaaa

File: source/sql/dwh/bq_view/L1/visibility/lgd_users.sql [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+      IF(REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.mainAddressId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.mainAddressId'), '"', '')) AS main_address_id,
+      D.main_address,
+      IF(REPLACE(JSON_EXTRACT_SCALAR(A.data, '$.usermetadata.assignedTaskId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT_SCALAR(A.data, '$.usermetadata.assignedTaskId'), '"', ''))  AS assigned_task_id,
+      JSON_EXTRACT_ARRAY(A.data, '$.usermetadata.apps') AS apps,
+      B.store_images,
+      IF(REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.userId'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.userId'), '"', '')) AS user_id,
+      IF(REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.storeCode'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.storeCode'), '"', '')) AS store_code,
+      G.integrated_account,
+      E.metadata,
+      IF(REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.approvedAt'), '"', '') = "", NULL, REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.approvedAt'), '"', '')) AS approved_at,
```
Comment (toileto) [inline source/sql/dwh/bq_view/L1/visibility/lgd_users.sql:225]: Kalau dari namanya, ini hasilnya harus di cast ke TIMESTAMP kayanya

Jadi:

```IF(REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.approvedAt'), '"', '') = "", NULL, CAST(REPLACE(JSON_EXTRACT(A.data, '$.usermetadata.approvedAt'), '"', '') AS TIMESTAMP))```

**Review by toileto — CHANGES_REQUESTED** [review]
Ada sedikit comment yaaaaa

**Review by toileto — APPROVED** [review]
Approved

**Pull request merged**